tokendito/duo_helpers.py

```diff
@@ -0,0 +1,274 @@
+            continue
+
+        if mfa_result.lower() == "success":
+            logging.debug("Successful MFA challenge received")
+            break
+        if mfa_result.lower() == "failure":
+            print("MFA challenge has failed for reason:"
+                  " {}\nPlease try again.".format(mfa_result, verify_mfa["reason"]))
+            continue
+        else:
+            logging.debug("MFA challenge result: {}"
+                          "Reason: {}\n\n".format(mfa_result, verify_mfa["reason"]))
+            continue
+        time.sleep(1)
+    return verify_mfa
+
+
+def duo_factor_callback(duo_info, verify_mfa):
+    """Inform factor callback api of successful challenge.
+
+    This request seems to inform this factor's callback url
+    that the challenge process has been completed.
+
+    :param duo_info: dict of parameters for Duo.
+    :param verify_mfa: verified mfa challenge response from status api.
+    :return sig_response: required to sign final Duo callback request.
+    """
+    factor_callback_url = "https://{}{}".format(
+        duo_info["host"], verify_mfa["result_url"])
+    factor_callback = duo_api_post(factor_callback_url, payload={
+        "sid": duo_info["sid"]})
+    sig_response = "{}:{}".format(
+        factor_callback.json()["response"]["cookie"], duo_info["app_sig"])
+
+    logging.debug("Completed factor callback.")
+    return sig_response
+
+
+def authenticate_duo(selected_okta_factor):
+    """Accomplish MFA via Duo.
+
+    This is the main function that coordinates the Duo
+    multifactor fetching, presentation, selection, challenge,
+    and verification until making an Okta callback.
+
+    :param selected_okta_factor: Duo factor information retrieved from Okta.
+    :return payload: required payload for Okta callback
+    :return headers: required headers for Okta callback
+    """
+    duo_info = prepare_duo_info(selected_okta_factor)
+
+    # Collect devices, factors, auth params for Duo
+    duo_info, duo_auth_response = get_duo_sid(duo_info)
+    factor_options = get_duo_devices(duo_auth_response)
+    mfa_index = helpers.select_preferred_mfa_index(factor_options, duo=True)
+    mfa_option = factor_options[mfa_index]
+    logging.debug("Selected MFA is [{}]".format(mfa_option))
+
+    if mfa_option["factor"].lower() == "passcode":
+        passcode = helpers.collect_totp()
+    else:
+        passcode = None
+
+    txid = duo_mfa_challenge(duo_info, mfa_option, passcode)
+    verify_mfa = duo_mfa_verify(duo_info, txid)
+
+    # Make factor callback to Duo
+    sig_response = duo_factor_callback(duo_info, verify_mfa)
+
+    # Prepare for Okta callback
+    payload = helpers.prepare_payload(id=duo_info["factor_id"],
+                                      sig_response=sig_response,
+                                      stateToken=duo_info["state_token"])
+    headers = {}
+    headers["content-type"] = "application/json"
+    headers["accept"] = "application/json"
+
```

Comment:
Can any of these be verified for failure / accurate response status?

Reply:
all of these functions have response validation inside them. this is the "main" of the duo process so i was trying to make it easy to read the flow: challenge duo mfa, verify that the user completes the challenge, callback/update challenge status to the factor in duo